scripts/compaction_bounds.zig

```diff
@@ -0,0 +1,128 @@
+                    while (table_count_remaining > 0) {
+                        level_count += 1;
+                        table_count_remaining = table_count_remaining -| tables_per_level;
+                        tables_per_level *= config.lsm_growth_factor;
+                    }
+                }
+
+                // Write bandwidth for flushing the immutable table:
+                bandwidth_bytes_per_second +=
+                    // Total size of all values
+                    transfers_per_second * @sizeOf(Value) *
+                    // All trees except the object tree might have to write tombstones
+                    @as(usize, if (std.meta.eql(field_name, "timestamp")) 1 else 2);
+
+                // Read/write bandwidth for compaction:
+                const bandwidth_bytes_per_compaction =
+                    // Compaction touches:
+                    // * One table from level a.
+                    // * Up to config.lsm_growth_factor tables from level b.
+                    // TODO Compaction for immutable table to level 0 doesn't have to read immutable table from disk.
+                    (1 + config.lsm_growth_factor) *
+                    // The data in each table must be read from disk and written back out.
+                    // TODO This factor could actually be as high as 4 with tombstones.
+                    2 * config.lsm_table_size_max;
+                // The object tree does not need to be compacted.
+                if (!std.meta.eql(field_name, "timestamp")) {
+                    bandwidth_bytes_per_second +=
+                        // In the ideal case we flush the immutable table when it contains exactly `values_per_table` transfers.
+                        (try std.math.divCeil(usize, transfers_per_second, values_per_table)) *
+                        // For every table ingested at level 0, we have to compact into each level once.
+                        level_count *
+                        bandwidth_bytes_per_compaction;
+                    switch (case) {
+                        .ideal => {},
+                        .actual => bandwidth_bytes_per_second +=
+                            // Currently, we flush every `config.lsm_batch_multiple` ops, causing many extra level 0 compactions.
+                            (try std.math.divCeil(usize, transfers_per_second, 8191 * config.lsm_batch_multiple)) *
+                            bandwidth_bytes_per_compaction,
+                        .actual_with_small_level0 => bandwidth_bytes_per_second +=
+                            // Currently, we flush every `config.lsm_batch_multiple` ops, causing many extra level 0 compactions.
+                            (try std.math.divCeil(usize, transfers_per_second, 8191 * config.lsm_batch_multiple)) *
+                            // With only one table in level 0, the extra compactions would be cheaper.
+                            2 * config.lsm_table_size_max,
```

Comment:
Here, we only need to write, since the level 0 table is most likely hot in cache.

Reply:
Currently the grid cache can only hold at most two tables and we have 10+ trees, so I'd expect compaction to be cold. 

We probably also don't want to touch the cache for compaction writes either because it flushes anything that is hot for the read workload, which makes it even less likely that the result of the last compaction will be hot.

Reply:
For large clusters, where there is significant write bandwidth, the grid cache would be sized to be a few GiB.

Our replacement policy is scan-resistant, given that we use CLOCK Nth-chance, so it's not a problem (and in fact a good idea) to put compaction writes into the grid cache. Otherwise, compactions of hot data which are in cache would leave the hot data out of the cache once compacted to the next level.

Reply:
:thinking: 

We'd need something like `tree_count * level_count * disk_table_size` bytes to cache the writes from the last round of compactions.